Libplanet/Net/Swarm.cs

```diff
@@ -611,12 +628,98 @@ CancellationToken cancellationToken
                         break;
                     }
 
+                case BlockHashes blockHashes:
+                    {
+                        await ProcessBlockHashes(
+                            blockHashes, blockchain, cancellationToken);
+                        break;
+                    }
+
                 default:
                     Trace.Fail($"Can't handle message. [{message}]");
                     break;
             }
         }
 
+        private async Task ProcessBlockHashes<T>(
+            BlockHashes message,
+            Blockchain<T> blockchain,
+            CancellationToken cancellationToken = default(CancellationToken))
+            where T : IAction
+        {
+            if (!(message.Identity is Address from))
+            {
+                throw new NullReferenceException(
+                    "BlockHashes doesn't have sender address.");
+            }
+
+            if (!_dealers.TryGetValue(from, out DealerSocket sock))
+            {
+                _logger.Information(
+                    "BlockHashes was sent from unknown peer. ignored.");
+                return;
+            }
+
+            IAsyncEnumerable<Block<T>> fetched = GetBlocksAsync<T>(
+                sock, message.Hashes, cancellationToken);
+
+            List<Block<T>> blocks = await fetched.ToListAsync();
+            await AppendBlocksAsync(
+                sock, blockchain, blocks, cancellationToken);
+            BlockReceived.Set();
+        }
+
+        private async Task AppendBlocksAsync<T>(
+            DealerSocket socket,
+            Blockchain<T> blockchain,
+            List<Block<T>> blocks,
+            CancellationToken cancellationToken)
+            where T : IAction
+        {
+            // We assume that the blocks are sorted in order.
+            Block<T> oldest = blocks.First();
+            Block<T> newest = blocks.Last();
+            HashDigest<SHA256>? tip = blockchain.Store.IndexBlockHash(-1);
+
+            if (tip == null || oldest.PreviousHash == tip)
```

Comment:
Does it mean the `newest` is expected to be the most closest to the genesis, unlike the intuition its name give us?

Reply:
No, `newest` is most far from genesis. 

This condition means that if `oldest` in received blocks is next to our tip, we can connect it directly.